libnopegl/src/node_colorstats.c

```diff
@@ -0,0 +1,414 @@
+    if (ret < 0)
+        return ret;
+
+    s->depth_index         = ngli_pgcraft_get_uniform_index(s->crafter_init, "depth",         NGLI_PROGRAM_SHADER_COMP);
+    s->length_minus1_index = ngli_pgcraft_get_uniform_index(s->crafter_init, "length_minus1", NGLI_PROGRAM_SHADER_COMP);
+    return 0;
+}
+
+/* Phase 2: compute waveform in the data field (histograms per column) */
+static int setup_waveform_compute(struct colorstats_priv *s, const struct pgcraft_block *block,
+                          const struct ngl_node *texture_node)
+{
+    struct texture_priv *texture_priv = texture_node->priv_data;
+    const struct texture_opts *texture_opts = texture_node->opts;
+    struct pgcraft_texture textures[] = {
+        {
+            .name        = "source",
+            .stage       = NGLI_PROGRAM_SHADER_COMP,
+            .image       = &texture_priv->image,
+            .format      = texture_priv->params.format,
+            .clamp_video = 0, /* clamping is done manually in the shader */
+        },
+    };
+
+    if (texture_opts->data_src && texture_opts->data_src->cls->id == NGL_NODE_MEDIA)
+        textures[0].type = NGLI_PGCRAFT_SHADER_TEX_TYPE_VIDEO;
+    else
+        textures[0].type = NGLI_PGCRAFT_SHADER_TEX_TYPE_2D;
+
+    const struct pgcraft_params crafter_params = {
+        .comp_base      = colorstats_waveform_comp,
+        .textures       = textures,
+        .nb_textures    = NGLI_ARRAY_NB(textures),
+        .blocks         = block,
+        .nb_blocks      = 1,
+        .workgroup_size = {s->group_size, 1, 1},
+    };
+
+    return setup_compute(s, s->crafter_waveform, s->pipeline_compat_waveform, &crafter_params);
+}
+
+/* Phase 3: summary and scale for global histograms */
+static int setup_sumscale_compute(struct colorstats_priv *s, const struct pgcraft_block *block)
+{
+    const struct pgcraft_params crafter_params = {
+        .comp_base      = colorstats_sumscale_comp,
+        .blocks         = block,
+        .nb_blocks      = 1,
+        .workgroup_size = {s->group_size, 1, 1},
+    };
+
+    return setup_compute(s, s->crafter_sumscale, s->pipeline_compat_sumscale, &crafter_params);
+}
+
+static int init_computes(struct ngl_node *node)
+{
+    struct ngl_ctx *ctx = node->ctx;
+    struct colorstats_priv *s = node->priv_data;
+    const struct colorstats_opts *o = node->opts;
+    struct gpu_ctx *gpu_ctx = ctx->gpu_ctx;
+
+    /*
+     * Define workgroup size using limits. We pick a value multiple of the
+     * depth on purpose.
+     *
+     * GL_MAX_COMPUTE_WORK_GROUP_SIZE must be at least [128,128,64] (OpenGLES
+     * 3.1), but GL_MAX_COMPUTE_WORK_GROUP_INVOCATIONS (x*y*z) is only 128.
+     * Meaning that if we pick a work group size X=128, we will have to use Y=1
+     * and Z=1.
+     */
+    const struct gpu_limits *limits = &gpu_ctx->limits;
+    const int max_group_size_x = limits->max_compute_work_group_size[0];
+    ngli_assert(max_group_size_x >= 128);
```

Comment:
Is this check really useful ? Is it to prevent bad drivers from running the code ? If so why not check the other limits ?

Reply:
It's a bit redundant with the comment, but since the line below is `s->group_size = max_group_size_x >= 256 ? 256 : 128;` I wanted to show the developer that 128 is an always safe fallback value. Do you want me to remove it?

Reply:
I'm in favor of removing it.

Reply:
okay, removed